src/app/clusters/thermostat-server/thermostat-server-atomic.cpp

```diff
@@ -0,0 +1,548 @@
+/// @param endpoint The associated endpoint for the AtomicRequest invocation
+/// @param attributeRequests The list of requested attributes
+/// @param attributeStatusCount The number of attribute statuses in attributeStatuses
+/// @param attributeStatuses The status of each requested attribute, plus additional attributes if needed
+/// @param requireAll Whether the caller requires all atomic attributes to be represented in attributeRequests
+/// @return Status::Success if the request is valid, an error status if it is not
+Status buildAttributeStatuses(const EndpointId endpoint, const DataModel::DecodableList<chip::AttributeId> attributeRequests,
+                              size_t & attributeStatusCount,
+                              Platform::ScopedMemoryBuffer<AtomicAttributeStatusStruct::Type> & attributeStatuses, bool requireAll)
+{
+
+    bool requestedPresets = false, requestedSchedules = false;
+    attributeStatusCount = 0;
+    if (!countAttributeRequests(attributeRequests, attributeStatusCount, requestedPresets, requestedSchedules))
+    {
+        // Either we errored reading the list, or one of the attributes is not supported on this server
+        return Status::InvalidCommand;
+    }
+    if (attributeStatusCount == 0)
+    {
+        // List can't be empty
+        return Status::InvalidCommand;
+    }
+    if (requestedPresets ^ requestedSchedules)
+    {
+        // Client requested presets or schedules, but not both, so we need an extra status
+        attributeStatusCount++;
+    }
+    attributeStatuses.Alloc(attributeStatusCount);
+    for (size_t i = 0; i < attributeStatusCount; ++i)
+    {
+        attributeStatuses[i].attributeID = kInvalidAttributeId;
+        attributeStatuses[i].statusCode  = 0;
+    }
+    auto attributeIdsIter = attributeRequests.begin();
+    size_t index          = 0;
+    while (attributeIdsIter.Next())
+    {
+        auto & attributeId = attributeIdsIter.GetValue();
+
+        for (size_t i = 0; i < index; ++i)
+        {
+            auto & attributeStatus = attributeStatuses[i];
+            if (attributeStatus.attributeID == attributeId)
+            {
+                // Double-requesting an attribute is invalid
+                return Status::InvalidCommand;
+            }
+        }
+        attributeStatuses[index].attributeID = attributeId;
+        attributeStatuses[index].statusCode  = to_underlying(Status::Success);
+        index++;
+    }
+    if (attributeIdsIter.GetStatus() != CHIP_NO_ERROR)
+    {
+        return Status::InvalidCommand;
+    }
+    for (size_t i = 0; i < index; ++i)
+    {
+        auto & attributeStatus = attributeStatuses[i];
+        const EmberAfAttributeMetadata * metadata =
+            emberAfLocateAttributeMetadata(endpoint, Thermostat::Id, attributeStatus.attributeID);
+
+        if (metadata == nullptr)
+        {
+            // This is not a valid attribute on the Thermostat cluster on the supplied endpoint
```

Comment:
What's the proper way to ask ember if an attribute is supported? 

I think the spec is fine; it says: if an attribute isn't supported at all, then return invalid command for the whole thing; if it's supported but doesn't support atomic writes, then return an attribute status to that effect. 

So the only problem here is that there are supported attributes without metadata, so I need some other way of checking.

Reply:
> What's the proper way to ask ember if an attribute is supported?

Your way is the right way _except_ for global attributes.  Those are supported on all clusters, by definition, so we don't have a bunch of in-memory metadata for them.

You're right that the spec defines a reasonable behavior here, so to implement that what we should do is:

1.  Check for metadata as you do now.
2. If no metadata, check whether the attribute ID is in `GlobalAttributesNotInMetadata` (see `src/app/GlobalAttributes.h`).  If it's in there, it's a supported attribute.  Otherwise it's not.